atintegrators/gwig.c

```diff
@@ -0,0 +1,305 @@
+  }
+}
+
+
+void GWigPass_2nd(struct gwig *pWig, double *X) 
+{
+  int    i, Nstep;
+  double dl;
+  
+  Nstep = pWig->PN*(pWig->Nw);
+  dl    = pWig->Lw/(pWig->PN);
+
+  for (i = 1; i <= Nstep; i++) {
+    GWigMap_2nd(pWig, X, dl);
+  }
+}
+
+
+void GWigPass_4th(struct gwig *pWig, double *X)
+{
+
+  const double x1 = 1.3512071919596576340476878089715e0;
+  const double x0 =-1.7024143839193152680953756179429e0;
+
+  int    i, Nstep;
+  double dl, dl1, dl0;
+ 
+  Nstep = pWig->PN*(pWig->Nw);
+  dl = pWig->Lw/(pWig->PN);
+
+  dl1 = x1*dl;
+  dl0 = x0*dl;
+
+  for (i = 1; i <= Nstep; i++ ) {
+    GWigMap_2nd(pWig, X, dl1);
+    GWigMap_2nd(pWig, X, dl0);
+    GWigMap_2nd(pWig, X, dl1);
+  }
+}
+
+
+void GWigMap_2nd(struct gwig *pWig, double *X, double dl) 
+{
+
+  double dld, dl2, dl2d;
+  double ax, ay, axpy, aypx;
+  
+  dld  = dl/(1.0e0 + X[4]);
+  dl2  = 0.5e0 * dl;
+  dl2d = dl2/(1.0e0 + X[4]);
+
+  /* Step1: increase a half step in z */
+  pWig->Zw = pWig->Zw + dl2;
+
+  /* Step2: a half drift in y */
+  GWigAy(pWig, X, &ay, &aypx);
+  X[1] = X[1] - aypx;
+  X[3] = X[3] - ay;
+
+  X[2] = X[2] + dl2d*X[3];
+  X[5] = X[5] + 0.5e0*dl2d*(X[3]*X[3])/(1.0e0+X[4]);
+   
+  GWigAy(pWig, X, &ay, &aypx);
+  X[1] = X[1] + aypx;
+  X[3] = X[3] + ay;
+
+  /* Step3: a full drift in x */
+  GWigAx(pWig, X, &ax, &axpy);
+  X[1] = X[1] - ax;
+  X[3] = X[3] - axpy;
+
+  X[0] = X[0] + dld*X[1];
+/* Full path length
+  X[5] = X[5] + dl + 0.5e0*dld*(X[1]*X[1])/(1.0e0+X[4]);
+*/
+  /* Differential path length only */
+  X[5] = X[5] + 0.5e0*dld*(X[1]*X[1])/(1.0e0+X[4]);
+   
+  GWigAx(pWig, X, &ax, &axpy);
+  X[1] = X[1] + ax;
+  X[3] = X[3] + axpy;
+
+  /* Step4: a half drift in y */
+  GWigAy(pWig, X, &ay, &aypx);
+  X[1] = X[1] - aypx;
+  X[3] = X[3] - ay;
+
+  X[2] = X[2] + dl2d*X[3];
+  X[5] = X[5] + 0.5e0*dl2d*(X[3]*X[3])/(1.0e0+X[4]);
+   
+  GWigAy(pWig, X, &ay, &aypx);
+  X[1] = X[1] + aypx;
+  X[3] = X[3] + ay;
+
+  /* Step5: increase a half step in z */
+  pWig->Zw = pWig->Zw + dl2;
+  
+}
+
+
+void GWigAx(struct gwig *pWig, double *Xvec, double *pax, double *paxpy) 
+{
+
+  int    i;
+  double x, y, z;
+  double kx, ky, kz, tz, kw;
+  double cx, sxkx, chx, shx;
+  double cy, sy, chy, shy, sz;
+  double gamma0, beta0;
+  double ax, axpy;
+
+  x = Xvec[0];
+  y = Xvec[2];
+  z = pWig->Zw;
+  
+  kw   = 2e0*PI/(pWig->Lw);
+  ax   = 0e0;
+  axpy = 0e0;
+
+  gamma0   = pWig->E0/XMC2;
+  beta0    = sqrt(1e0 - 1e0/(gamma0*gamma0));
+  pWig->Aw = (q_e/m_e/clight)/(2e0*PI) * (pWig->Lw) * (pWig->PB0);
+
+  /* Horizontal Wiggler: note that one potentially could have: kx=0 */
+  for (i = 0; i < pWig->NHharm; i++) {
+    pWig->HCw[i] = pWig->HCw_raw[i]*(pWig->Aw)/(gamma0*beta0);
+    kx = pWig->Hkx[i];
+    ky = pWig->Hky[i];
+    kz = pWig->Hkz[i];
+    tz = pWig->Htz[i];
+
+    cx  = cos(kx*x);
+    chy = cosh(ky * y);
+    sz  = sin(kz * z + tz);
+    ax  = ax + (pWig->HCw[i])*(kw/kz)*cx*chy*sz;
+
+    shy = sinh(ky * y);
+    if ( abs(kx/kw) > GWIG_EPS ) {

```

Comment:
I still get a warning (justified according to me) about using `abs` instead of `fabs`:

```
/Volumes/machfs/machine/at/atintegrators/gwig.c:196:10: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
    if ( abs(kx/kw) > GWIG_EPS ) {
         ^
/Volumes/machfs/machine/at/atintegrators/gwig.c:196:10: note: use function 'fabs' instead
    if ( abs(kx/kw) > GWIG_EPS ) {
         ^~~
         fabs
/Volumes/machfs/machine/at/atintegrators/gwig.c:284:9: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]
    if (abs(ky/kw) > GWIG_EPS) {
        ^
/Volumes/machfs/machine/at/atintegrators/gwig.c:284:9: note: use function 'fabs' instead
    if (abs(ky/kw) > GWIG_EPS) {
        ^~~
        fabs

```
Same comment for line 284